File: src/cli/terminal.ts
/**
 * Terminal display — 与 Python terminal.py 对应
 * thinking 深灰 / answer 亮色 / 长思考折叠 / 多轮分隔 / 步骤编号 / 代码写入打字机效果
 */

// 代码写入流式显示的配置
const CODE_STREAM_CONFIG = {
  // 小文件 (≤50行): 每行延迟 5ms
  smallLineDelay: 5,
  smallThreshold: 50,
  // 中文件 (≤200行): 每行延迟 2ms
  mediumLineDelay: 2,
  mediumThreshold: 200,
  // 大文件 (>200行): 只显示首尾，每行延迟 1ms
  largeLineDelay: 1,
  largeHeadLines: 30,
  largeTailLines: 10,
  // 最小内容长度（太短不流式）
  minLength: 30,
};

export class Terminal {
  private buf: string[] = [];
  private shownReasoning = false;
  private showingAnswer = false;
  private round = 0;
  private step = 0;
  private codeStreamEnabled = true;

  static readonly DEEP  = "\x1b[38;5;239m";
  static readonly CYAN  = "\x1b[38;5;51m";
  static readonly GREEN = "\x1b[38;5;82m";
  static readonly YELLOW= "\x1b[38;5;220m";
  static readonly RED   = "\x1b[38;5;196m";
  static readonly GRAY  = "\x1b[38;5;240m";
  static readonly DIM   = "\x1b[38;5;245m";
  static readonly BOLD  = "\x1b[1m";
  static readonly RESET = "\x1b[0m";
  static readonly FOLD_LEN = 200;
  static readonly FOLD_PREVIEW = 80;

  write(s: string) { process.stdout.write(s); }

  // Think phase — deep grey
  thinkToken(token: string) {
    if (!this.shownReasoning) {
      this.shownReasoning = true;
      this.write(`\n${Terminal.DEEP}`);
    }
    this.buf.push(token);
    this.write(token);
  }

  // Answer phase — bright with separator
  answerToken(token: string) {
    if (!this.showingAnswer) {
      if (this.shownReasoning) {
        const reasoning = this.buf.join("");
        this.write(Terminal.RESET);
        if (reasoning.length > Terminal.FOLD_LEN) {
          const flat = reasoning.replace(/\n/g, " ").trim();
          const preview = flat.length > Terminal.FOLD_PREVIEW
            ? flat.slice(0, Terminal.FOLD_PREVIEW - 3) + "..."
            : flat;
          this.write(`\n  ${Terminal.DIM}💭 ${preview}${Terminal.RESET}\n\n`);
        } else {
          this.write("\n");
        }
      }
      this.buf = [];
      this.showingAnswer = true;
    }
    this.write(token);
  }

  // Tool call — step number + name + params
  toolStart(name: string, args: Record<string, unknown>) {
    this.step++;
    const argsStr = fmtArgs(args);
    if (argsStr) {
      this.write(`\n  ${Terminal.GRAY}[${this.step}]${Terminal.RESET} ${Terminal.CYAN}▸ ${name}${Terminal.RESET} ${Terminal.DIM}(${argsStr})${Terminal.RESET}`);
    } else {
      this.write(`\n  ${Terminal.GRAY}[${this.step}]${Terminal.RESET} ${Terminal.CYAN}▸ ${name}${Terminal.RESET}`);
    }
  }

  toolDone(success: boolean, latencyMs: number, preview: string) {
    const icon = success ? `${Terminal.GREEN}✓${Terminal.RESET}` : `${Terminal.RED}✗${Terminal.RESET}`;
    const short = preview.replace(/\n/g, " ").trim().slice(0, 80);
    if (short) {
      this.write(` ${icon} ${Terminal.GRAY}[${latencyMs.toFixed(0)}ms]${Terminal.RESET} ${Terminal.DIM}${short}${Terminal.RESET}\n`);
    } else {
      this.write(` ${icon} ${Terminal.GRAY}[${latencyMs.toFixed(0)}ms]${Terminal.RESET}\n`);
    }
  }

  // ── 子代理进度显示 ──

  /** 子代理派遣开始 — 显示总数 */
  subagentDispatch(count: number) {
    const label = count === 1 ? "派遣子代理..." : `派遣 ${count} 个子代理...`;
    this.write(`\n  ${Terminal.YELLOW}⚡ ${label}${Terminal.RESET}\n`);
  }

  /** 单个子代理开始执行 */
  subagentStart(idx: number, total: number, task: string) {
    const preview = task.length > 50 ? task.slice(0, 47) + "..." : task;
    this.write(`  ${Terminal.GRAY}  └ [${idx}/${total}]${Terminal.RESET} ${Terminal.DIM}▶ ${preview}${Terminal.RESET}\n`);
  }

  /** 单个子代理完成 */
  subagentDone(idx: number, total: number, success: boolean, latencyMs?: number) {
    const icon = success ? `${Terminal.GREEN}✓${Terminal.RESET}` : `${Terminal.RED}✗${Terminal.RESET}`;
    const time = latencyMs != null ? ` ${Terminal.GRAY}[${(latencyMs / 1000).toFixed(1)}s]${Terminal.RESET}` : "";
    this.write(`  ${Terminal.GRAY}  └ [${idx}/${total}]${Terminal.RESET} ${icon}${time}\n`);
  }

  /** Check if answer text has been displayed to the user */
  isAnswerShown(): boolean {
    return this.showingAnswer;
  }

  /** Explicitly write answer text to terminal (fallback when streaming didn't work) */
  writeAnswer(text: string) {
    if (!this.showingAnswer) {
      this.closeThinking();
      this.showingAnswer = true;
    }
    this.write(text);
  }

  closeThinking() {
    if (this.shownReasoning && !this.showingAnswer) {
      const reasoning = this.buf.join("");
      this.write(Terminal.RESET);
      if (reasoning.length > Terminal.FOLD_LEN) {
        const flat = reasoning.replace(/\n/g, " ").trim();
        const preview = flat.length > Terminal.FOLD_PREVIEW
          ? flat.slice(0, Terminal.FOLD_PREVIEW - 3) + "..."
          : flat;
        this.write(`\n  ${Terminal.DIM}💭 ${preview}${Terminal.RESET}\n`);
      } else {
        this.write("\n");
      }
      this.buf = [];
    }
  }

  // ── 代码写入打字机效果 ──

  /** 启用/禁用代码流式显示 */
  setCodeStream(enabled: boolean) { this.codeStreamEnabled = enabled; }

  /**
   * 流式显示代码写入过程 — 打字机效果。
   * 
   * 策略:
   *   - 小文件 (≤50行): 全部显示，每行 5ms 延迟
   *   - 中文件 (≤200行): 全部显示，每行 2ms 延迟
   *   - 大文件 (>200行): 首尾显示 + 中间省略，每行 1ms 延迟
   *   - 极短内容 (<30字符): 不流式，直接显示
   */
  async codeStream(filePath: string, content: string): Promise<void> {
    if (!this.codeStreamEnabled || !content || content.length < CODE_STREAM_CONFIG.minLength) return;

    const fileName = filePath.split(/[/\\]/).pop() || filePath;
    const ext = fileName.split(".").pop()?.toLowerCase() || "";
    const lines = content.split("\n");
    const totalLines = lines.length;
    const totalChars = content.length;

    // 代码颜色: 根据文件类型选色
    const codeColor = this._codeColor(ext);

    // 头部: 文件名 + 行数
    this.write(`\n  ${Terminal.DIM}✎ ${Terminal.RESET}${Terminal.CYAN}${fileName}${Terminal.RESET} ${Terminal.GRAY}(${totalLines} 行, ${totalChars.toLocaleString()} 字符)${Terminal.RESET}\n`);
    this.write(`  ${Terminal.GRAY}┌${"─".repeat(Math.min(fileName.length + 20, 60))}┐${Terminal.RESET}\n`);

    const cfg = CODE_STREAM_CONFIG;
    let displayLines: string[];
    let lineDelay: number;
    let omitted = 0;

    if (totalLines <= cfg.smallThreshold) {
      displayLines = lines;
      lineDelay = cfg.smallLineDelay;
    } else if (totalLines <= cfg.mediumThreshold) {
      displayLines = lines;
      lineDelay = cfg.mediumLineDelay;
    } else {
      // 大文件: 首尾显示
      const head = lines.slice(0, cfg.largeHeadLines);
      const tail = lines.slice(-cfg.largeTailLines);
      omitted = totalLines - cfg.largeHeadLines - cfg.largeTailLines;
      displayLines = [...head, `__OMITTED__${omitted}__`, ...tail];
      lineDelay = cfg.largeLineDelay;
    }

    // 流式输出每一行
    for (const line of displayLines) {
      if (line.startsWith("__OMITTED__")) {
        const count = parseInt(line.match(/\d+/)?.[0] || "0");
        this.write(`  ${Terminal.GRAY}│  ... 省略 ${count} 行 ...${Terminal.RESET}\n`);
      } else {
        // 截断超长行
        const displayLine = line.length > 120 ? line.slice(0, 117) + "..." : line;
        this.write(`  ${Terminal.GRAY}│${Terminal.RESET} ${codeColor}${displayLine}${Terminal.RESET}\n`);
      }
      if (lineDelay > 0) {
        await new Promise(r => setTimeout(r, lineDelay));
      }
    }

    // 底部
    this.write(`  ${Terminal.GRAY}└${"─".repeat(Math.min(fileName.length + 20, 60))}┘${Terminal.RESET}\n`);
  }

  /** 根据文件扩展名返回代码颜色 */
  private _codeColor(ext: string): string {
    const map: Record<string, string> = {
      ts: "\x1b[38;5;75m",   // 蓝
      tsx: "\x1b[38;5;75m",
      js: "\x1b[38;5;221m",  // 黄
      jsx: "\x1b[38;5;221m",
      py: "\x1b[38;5;114m",  // 绿
      html: "\x1b[38;5;209m", // 橙
      css: "\x1b[38;5;141m",  // 紫
      json: "\x1b[38;5;215m", // 浅橙
      md: "\x1b[38;5;250m",   // 浅灰
      yml: "\x1b[38;5;215m",
      yaml: "\x1b[38;5;215m",
      sql: "\x1b[38;5;117m",
      sh: "\x1b[38;5;114m",
      go: "\x1b[38;5;81m",
      rs: "\x1b[38;5;173m",
      vue: "\x1b[38;5;114m",
    };
    return map[ext] || Terminal.DIM;
  }

  nextRound() {
    this.round++;
    this.step = 0;
    this.buf = [];
    this.shownReasoning = false;
    this.showingAnswer = false;
    // Multi-round separator (skip first round)
    if (this.round > 1) {
      this.write(`\n  ${Terminal.GRAY}${"─".repeat(44)}${Terminal.RESET}\n`);
    }
  }

  // 权限模式元数据
  private static readonly MODE_META: Record<string, {color: string; icon: string; label: string; desc: string}> = {
    standard: { color: Terminal.GREEN,  icon: "🛡", label: "Standard", desc: "安全模式" },
    auto:     { color: Terminal.YELLOW, icon: "✎", label: "Auto",     desc: "自动模式" },
    yolo:     { color: Terminal.RED,    icon: "⚠", label: "YOLO",    desc: "无限制" },
  };

  banner(model: string, tools: number, workDir: string, mode: string, sessionId?: string, contextLimit?: number, isResume?: boolean) {
    const meta = Terminal.MODE_META[mode] || { color: Terminal.GRAY, icon: "?", label: mode, desc: "" };
    // 格式化上下文容量
    let ctxStr = "";
    if (contextLimit && contextLimit > 0) {
      ctxStr = contextLimit >= 1_000_000
        ? `${Math.floor(contextLimit / 1_000_000)}M ctx`
        : `${Math.floor(contextLimit / 1000)}K ctx`;
    }
    this.write(`\n${Terminal.CYAN}╔${"═".repeat(52)}╗${Terminal.RESET}\n`);
    // 模型行
    let modelLine = `  ${Terminal.BOLD}Cortex Agent${Terminal.RESET}  ${Terminal.GREEN}${model}${Terminal.RESET}`;
    if (ctxStr) modelLine += `  ${Terminal.GRAY}${ctxStr}${Terminal.RESET}`;
    modelLine += `  ${Terminal.GRAY}${tools} tools  🟦${Terminal.RESET}`;
    this.write(modelLine + "\n");
    // 权限行
    this.write(`  ${meta.color}${meta.icon} ${meta.label}${Terminal.RESET}  ${Terminal.DIM}${meta.desc}${Terminal.RESET}  ${Terminal.GRAY}(Shift+Tab 切换)${Terminal.RESET}\n`);
    if (sessionId) this.write(`  ${Terminal.GRAY}Session: ${sessionId}${isResume ? " (已恢复)" : " (新会话)"}${Terminal.RESET}\n`);
    this.write(`  ${Terminal.GRAY}${workDir}${Terminal.RESET}\n`);
    this.write(`${Terminal.CYAN}╚${"═".repeat(52)}╝${Terminal.RESET}\n`);
  }

  error(msg: string) {
    this.write(`\n  ${Terminal.RED}✗ ${msg}${Terminal.RESET}\n`);
  }
}

function fmtArgs(args: Record<string, unknown>): string {
  const parts: string[] = [];
  for (const [k, v] of Object.entries(args)) {
    if (k === "workDir" || k === "work_dir") continue;
    let s = String(v);
    if (s.length > 50) s = s.slice(0, 47) + "...";
    parts.push(`${k}=${s}`);
    if (parts.length >= 4) break; // max 4 params shown
  }
  return parts.join(", ");
}

File: python/cortex_agent/terminal.py
"""
Cortex Agent 终端显示器 — Harness Agent Think → Guard → Act → Reflect 差异化输出

设计哲学：
  Think 阶段 — 深灰流式（可见、差异化），长思考自动折叠为一行摘要
  Act 阶段  — 青色工具标签 + 参数摘要 + 灰色结果
  Answer    — 亮色输出，与思考阶段有清晰视觉分隔
  多轮对话  — 每轮之间有分隔线 + 步骤编号
"""

import sys
from typing import List


class Terminal:
    """终端流式显示：thinking deep-grey, answer bright, long-thinking fold, code typewriter"""

    DEEP  = "\033[38;5;239m"
    CYAN  = "\033[38;5;51m"
    GREEN = "\033[38;5;82m"
    YELLOW= "\033[38;5;220m"
    RED   = "\033[38;5;196m"
    GRAY  = "\033[38;5;240m"
    DIM   = "\033[38;5;245m"
    BOLD  = "\033[1m"
    RESET = "\033[0m"

    # 折叠阈值：超过任一阈值即视为长思考
    FOLD_CHAR_THRESHOLD = 200
    FOLD_LINE_THRESHOLD = 3
    FOLD_PREVIEW_LEN    = 80

    # ── 代码写入流式显示配置 ──
    CODE_SMALL_LINE_DELAY = 0.005   # 小文件 (≤50行): 每行 5ms
    CODE_SMALL_THRESHOLD  = 50
    CODE_MEDIUM_LINE_DELAY = 0.002  # 中文件 (≤200行): 每行 2ms
    CODE_MEDIUM_THRESHOLD  = 200
    CODE_LARGE_LINE_DELAY  = 0.001  # 大文件 (>200行): 每行 1ms
    CODE_LARGE_HEAD_LINES  = 30
    CODE_LARGE_TAIL_LINES  = 10
    CODE_MIN_LENGTH        = 30     # 极短内容不流式

    # 代码颜色映射
    _CODE_COLORS = {
        "ts": "\033[38;5;75m", "tsx": "\033[38;5;75m",
        "js": "\033[38;5;221m", "jsx": "\033[38;5;221m",
        "py": "\033[38;5;114m",
        "html": "\033[38;5;209m", "css": "\033[38;5;141m",
        "json": "\033[38;5;215m", "md": "\033[38;5;250m",
        "yml": "\033[38;5;215m", "yaml": "\033[38;5;215m",
        "sql": "\033[38;5;117m", "sh": "\033[38;5;114m",
        "go": "\033[38;5;81m", "rs": "\033[38;5;173m",
        "vue": "\033[38;5;114m",
    }

    def __init__(self, enabled: bool = True):
        self.enabled = enabled
        self._buf: List[str] = []          # reasoning token buffer
        self._shown_reasoning = False       # reasoning was ever emitted this round
        self._showing_answer = False        # first answer token emitted this round
        self._round = 0                     # current round (increment per _loop)
        self._step = 0                      # current step within round (increment per _think)
        self._code_stream_enabled = True    # 代码写入打字机效果开关

    def _w(self, s: str):
        if self.enabled:
            sys.stdout.write(s)
            sys.stdout.flush()

    # ── Think phase ──

    def think_token(self, token: str):
        """Stream reasoning in deep grey — visible, differentiated."""
        if not self.enabled:
            return
        if not self._shown_reasoning:
            self._shown_reasoning = True
            self._w(f"\n{self.DEEP}")
        self._buf.append(token)
        self._w(token)

    # ── Transition helpers ──

    def _end_reasoning(self) -> str:
        """Close deep-grey colour. Returns full buffered reasoning text.
        Idempotent — safe to call multiple times."""
        if not self._shown_reasoning:
            return ""
        self._shown_reasoning = False
        self._w(self.RESET)
        return "".join(self._buf)

    def _is_short(self, text: str) -> bool:
        """Short reasoning → smooth transition; long → fold summary."""
        lines = text.count("\n") + 1
        return len(text) <= self.FOLD_CHAR_THRESHOLD and lines < self.FOLD_LINE_THRESHOLD

    def close_thinking(self):
        """Tool transition — close reasoning colour before tool labels appear."""
        reasoning = self._end_reasoning()
        if reasoning:
            # Long reasoning → fold to one-line summary before tools
            if not self._is_short(reasoning):
                flat = reasoning.replace("\n", " ").strip()
                preview = flat[:self.FOLD_PREVIEW_LEN]
                if len(flat) > self.FOLD_PREVIEW_LEN:
                    preview = preview[:self.FOLD_PREVIEW_LEN - 3] + "..."
                self._w(f"\n  {self.DIM}💭 {preview}{self.RESET}\n")
            else:
                self._w(self.RESET)
            self._w("\n")
            self._buf.clear()

    # ── Answer phase ──

    def answer_token(self, token: str):
        """Stream answer in default bright. First call transitions from thinking:
        - Short thinking: smooth colour reset + newline
        - Long thinking: collapsed one-line dim summary then newline"""
        if not self.enabled:
            return
        if not self._showing_answer:
            reasoning = self._end_reasoning()
            if reasoning and not self._is_short(reasoning):
                # Long reasoning → fold to one-line summary
                flat = reasoning.replace("\n", " ").strip()
                preview = flat[:self.FOLD_PREVIEW_LEN]
                if len(flat) > self.FOLD_PREVIEW_LEN:
                    preview = preview[:self.FOLD_PREVIEW_LEN - 3] + "..."
                self._w(f"\n  {self.DIM}💭 {preview}{self.RESET}\n\n")
            else:
                # Short or no reasoning → smooth transition
                self._w("\n")
            self._buf.clear()
            self._showing_answer = True
        self._w(token)

    # ── Round lifecycle ──

    def next_round(self):
        """Reset per-round state — called at start of each _loop iteration."""
        self._round += 1
        self._step = 0
        self._buf.clear()
        self._shown_reasoning = False
        self._showing_answer = False
        # Multi-round separator (skip first round)
        if self._round > 1:
            self._w(f"\n  {self.GRAY}{'─'*44}{self.RESET}\n")

    # ── Act phase (tools) ──

    def tool_start(self, name: str, args: dict):
        self._step += 1
        # Show step number + tool name + key params
        args_str = _fmt_args(args)
        if args_str:
            self._w(f"\n  {self.GRAY}[{self._step}]{self.RESET} {self.CYAN}▸ {name}{self.RESET} {self.DIM}({args_str}){self.RESET}")
        else:
            self._w(f"\n  {self.GRAY}[{self._step}]{self.RESET} {self.CYAN}▸ {name}{self.RESET}")

    def tool_inline(self, name: str, args: dict):
        pass

    def tool_done(self, success: bool, latency_ms: float, preview: str):
        icon = f"{self.GREEN}✓{self.RESET}" if success else f"{self.RED}✗{self.RESET}"
        short = preview.replace("\n", " ").strip()[:80]
        if short:
            self._w(f" {icon} {self.GRAY}[{latency_ms:.0f}ms]{self.RESET} {self.DIM}{short}{self.RESET}\n")
        else:
            self._w(f" {icon} {self.GRAY}[{latency_ms:.0f}ms]{self.RESET}\n")

    def tool_heartbeat(self, elapsed_seconds: int):
        """显示工具执行心跳（每 5 秒更新一次）"""
        self._w(f"  {self.GRAY}⏳ {elapsed_seconds}s...{self.RESET}\n")

    def tool_interrupted(self):
        """显示工具被用户中断"""
        self._w(f"  {self.YELLOW}⚠ 用户中断 (Ctrl+C){self.RESET}\n")

    # ── Banner ──

    # 权限模式元数据
    _MODE_META = {
        "standard": {"color": "\033[38;5;82m", "icon": "🛡", "label": "Standard", "desc": "安全模式"},
        "auto":     {"color": "\033[38;5;220m", "icon": "✎", "label": "Auto",     "desc": "自动模式"},
        "yolo":     {"color": "\033[38;5;196m", "icon": "⚠", "label": "YOLO",    "desc": "无限制"},
    }

    def banner(self, model: str, tools: int, work_dir: str, session_id: str = "", mode: str = "standard", context_limit: int = 0, is_resume: bool = False):
        meta = self._MODE_META.get(mode, {"color": self.GRAY, "icon": "?", "label": mode, "desc": ""})
        mc = meta["color"]
        mi = meta["icon"]
        ml = meta["label"]
        md = meta["desc"]
        # 格式化上下文容量
        ctx_str = ""
        if context_limit > 0:
            if context_limit >= 1_000_000:
                ctx_str = f"{context_limit // 1_000_000}M ctx"
            else:
                ctx_str = f"{context_limit // 1000}K ctx"
        # 顶边
        self._w(f"\n{self.CYAN}╔{'═'*52}╗{self.RESET}\n")
        # 模型行
        model_line = f"  {self.BOLD}Cortex Agent{self.RESET}  {self.GREEN}{model}{self.RESET}"
        if ctx_str:
            model_line += f"  {self.GRAY}{ctx_str}{self.RESET}"
        model_line += f"  {self.GRAY}{tools} tools  🐍{self.RESET}"
        self._w(model_line + "\n")
        # 权限行
        perm_line = f"  {mc}{mi} {ml}{self.RESET}  {self.DIM}{md}{self.RESET}  {self.GRAY}(/mode 切换){self.RESET}"
        self._w(perm_line + "\n")
        # Session
        if session_id:
            resume_tag = " (已恢复)" if is_resume else " (新会话)"
            self._w(f"  {self.GRAY}Session: {session_id}{resume_tag}{self.RESET}\n")
        # 工作目录
        self._w(f"  {self.GRAY}{work_dir}{self.RESET}\n")
        # 底边
        self._w(f"{self.CYAN}╚{'═'*52}╝{self.RESET}\n")

    def error(self, msg: str):
        self._w(f"\n  {self.RED}✗ {msg}{self.RESET}\n")

    # ── 代码写入打字机效果 ──

    def set_code_stream(self, enabled: bool):
        """启用/禁用代码流式显示"""
        self._code_stream_enabled = enabled

    def code_stream(self, file_path: str, content: str):
        """流式显示代码写入过程 — 打字机效果。
        
        策略:
          - 小文件 (≤50行): 全部显示，每行 5ms 延迟
          - 中文件 (≤200行): 全部显示，每行 2ms 延迟
          - 大文件 (>200行): 首尾显示 + 中间省略，每行 1ms 延迟
          - 极短内容 (<30字符): 不流式，直接跳过
        """
        import time, os
        if not self._code_stream_enabled or not content or len(content) < self.CODE_MIN_LENGTH:
            return

        file_name = os.path.basename(file_path) or file_path
        ext = file_name.rsplit(".", 1)[-1].lower() if "." in file_name else ""
        lines = content.split("\n")
        total_lines = len(lines)
        total_chars = len(content)
        code_color = self._CODE_COLORS.get(ext, self.DIM)

        # 头部
        border_len = min(len(file_name) + 20, 60)
        self._w(f"\n  {self.DIM}✎ {self.RESET}{self.CYAN}{file_name}{self.RESET} {self.GRAY}({total_lines} 行, {total_chars:,} 字符){self.RESET}\n")
        self._w(f"  {self.GRAY}┌{'─' * border_len}┐{self.RESET}\n")

        # 决定显示策略
        if total_lines <= self.CODE_SMALL_THRESHOLD:
            display_lines = lines
            line_delay = self.CODE_SMALL_LINE_DELAY
        elif total_lines <= self.CODE_MEDIUM_THRESHOLD:
            display_lines = lines
            line_delay = self.CODE_MEDIUM_LINE_DELAY
        else:
            head = lines[:self.CODE_LARGE_HEAD_LINES]
            tail = lines[-self.CODE_LARGE_TAIL_LINES:]
            omitted = total_lines - self.CODE_LARGE_HEAD_LINES - self.CODE_LARGE_TAIL_LINES
            display_lines = head + [f"__OMITTED__{omitted}__"] + tail
            line_delay = self.CODE_LARGE_LINE_DELAY

        # 流式输出
        for line in display_lines:
            if line.startswith("__OMITTED__"):
                count = int(line.replace("__OMITTED__", "").replace("__", ""))
                self._w(f"  {self.GRAY}│  ... 省略 {count} 行 ...{self.RESET}\n")
            else:
                display_line = line[:117] + "..." if len(line) > 120 else line
                self._w(f"  {self.GRAY}│{self.RESET} {code_color}{display_line}{self.RESET}\n")
            if line_delay > 0:
                time.sleep(line_delay)

        # 底部
        self._w(f"  {self.GRAY}└{'─' * border_len}┘{self.RESET}\n")


def _fmt_args(args: dict) -> str:
    """Format tool args for display — show key params concisely."""
    parts = []
    for k, v in args.items():
        if k in ("work_dir", "workDir"):
            continue
        s = str(v)
        if len(s) > 50:
            s = s[:47] + "..."
        # Quote string values
        if isinstance(v, str) and " " not in s and len(s) < 50:
            parts.append(f"{k}={s}")
        else:
            parts.append(f"{k}={s}")
    return ", ".join(parts[:4])  # max 4 params shown
